azure/services/tags/client.go

```diff
@@ -19,51 +19,68 @@ package tags
 import (
 	"context"
 
-	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-10-01/resources"
-	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+	"github.com/pkg/errors"
 	"sigs.k8s.io/cluster-api-provider-azure/azure"
 	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
 )
 
 // client wraps go-sdk.
 type client interface {
-	GetAtScope(context.Context, string) (resources.TagsResource, error)
-	UpdateAtScope(context.Context, string, resources.TagsPatchResource) (resources.TagsResource, error)
+	GetAtScope(context.Context, string) (armresources.TagsResource, error)
+	UpdateAtScope(context.Context, string, armresources.TagsPatchResource) (armresources.TagsResource, error)
 }
 
 // AzureClient contains the Azure go-sdk client.
 type AzureClient struct {
-	tags resources.TagsClient
+	tags *armresources.TagsClient
 }
 
 var _ client = (*AzureClient)(nil)
 
-// NewClient creates a new tags client from subscription ID.
-func NewClient(auth azure.Authorizer) *AzureClient {
-	c := newTagsClient(auth.SubscriptionID(), auth.BaseURI(), auth.Authorizer())
-	return &AzureClient{c}
-}
-
-// newTagsClient creates a new tags client from subscription ID.
-func newTagsClient(subscriptionID string, baseURI string, authorizer autorest.Authorizer) resources.TagsClient {
-	tagsClient := resources.NewTagsClientWithBaseURI(baseURI, subscriptionID)
-	azure.SetAutoRestClientDefaults(&tagsClient.Client, authorizer)
-	return tagsClient
+// NewClient creates a tags client from an authorizer.
+func NewClient(auth azure.Authorizer) (*AzureClient, error) {
+	opts, err := azure.ARMClientOptions(auth.CloudEnvironment())
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create tags client options")
+	}
+	factory, err := armresources.NewClientFactory(auth.SubscriptionID(), auth.Token(), opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create armresources client factory")
+	}
+	return &AzureClient{factory.NewTagsClient()}, nil
 }
 
 // GetAtScope sends the get at scope request.
-func (ac *AzureClient) GetAtScope(ctx context.Context, scope string) (resources.TagsResource, error) {
+func (ac *AzureClient) GetAtScope(ctx context.Context, scope string) (tagsResource armresources.TagsResource, err error) {
 	ctx, _, done := tele.StartSpanWithLogger(ctx, "tags.AzureClient.GetAtScope")
 	defer done()
 
-	return ac.tags.GetAtScope(ctx, scope)
+	// Handle a panic from the SDK by returning an error. This only happens in a managedcontrolplane unit test
+	// where the tags client is not mocked.
+	defer func() {
```

Comment:
I debugged the path of the `TestManagedControlPlaneScope_IsVnetManagedCache` test case "no Cache value", and it eventually leads here. In SDKv1 that led to returning an error when network communication failed, but in SDKv2 it panics.

It doesn't look easy to inject a mock into this code path, and there don't seem to be any accessible `nil` fields that we could fill in to placate the test environment. I'm not proud of it, but after several hours in the debugger I think this might be the expedient solution.

Reply:
With ASO we won't need this tags service at all, so I think relatively dirty hacks here are acceptable.

Reply:
^ Seeing now that this is used more widely than I first thought, I shouldn't be too sure about that but I'm still optimistic.

Reply:
not sure I love this either... I would rather us remove the test or change the test to make it work with this new code rather than hack the code just to make the unit test pass

Reply:
It seems like a valid test, and I've tried pretty hard to find a way to rework it. I can look again.

Reply:
I created a `VnetDescriber` interface so we could mock `virtualNetworksSvc.IsManaged()`. This client creates that service just-in-time, unless we pass one into the scope.

(What does mocking virtualnetworks in a managedcluster unit test have to do with refactoring the tags service? Nothing directly, but this short-circuits the code path before it gets to the failing tags call.)

No need to panic. :-) This isn't the cleanest change but it's better. Let me know what you think.

Reply:
Thank you, this is much better (and we're actually testing the thing the test intends to test).

Will let @nojnhuh chime in 

/lgtm

Reply:
+1, this approach lgtm.